src/Discord/Internal/Types/ScheduledEvents.hs

```diff
@@ -0,0 +1,188 @@
+    Object
+      $ (if isJust scheduledEventStageUserCount
+          then insert "user_count" (toJSON scheduledEventStageUserCount)
+          else id
+        )
+      $ (if isJust scheduledEventStageCreator
+          then insert "creator" (toJSON scheduledEventStageCreator)
+          else id
+        )
+          base
+   where
+    Object base = object
+      [ "id" .= scheduledEventStageId
+      , "guildId" .= scheduledEventStageGuildId
+      , "channel_id" .= scheduledEventStageChannelId
+      , "creator_id" .= scheduledEventStageCreatorId
+      , "name" .= scheduledEventStageName
+      , "description" .= scheduledEventStageDescription
+      , "scheduled_start_time" .= scheduledEventStageStartTime
+      , "scheduled_end_time" .= scheduledEventStageEndTime
+      , "privacy_level" .= scheduledEventStagePrivacyLevel
+      , "status" .= scheduledEventStageStatus
+      , "entity_type" .= Number 1
+      , "entity_id" .= scheduledEventStageEntityId
+      , "entity_metadata" .= Null
+      , "creator" .= scheduledEventStageCreator
+      , "user_count" .= scheduledEventStageUserCount
+      , "image" .= scheduledEventStageImage
+      ]
+  toJSON ScheduledEventVoice {..} =
+    Object
+      $ (if isJust scheduledEventVoiceUserCount
+          then insert "user_count" (toJSON scheduledEventVoiceUserCount)
+          else id
+        )
+      $ (if isJust scheduledEventVoiceCreator
+          then insert "creator" (toJSON scheduledEventVoiceCreator)
+          else id
+        )
+          base
+   where
+    Object base = object
+      [ "id" .= scheduledEventVoiceId
+      , "guildId" .= scheduledEventVoiceGuildId
+      , "channel_id" .= scheduledEventVoiceChannelId
+      , "creator_id" .= scheduledEventVoiceCreatorId
+      , "name" .= scheduledEventVoiceName
+      , "description" .= scheduledEventVoiceDescription
+      , "scheduled_start_time" .= scheduledEventVoiceStartTime
+      , "scheduled_end_time" .= scheduledEventVoiceEndTime
+      , "privacy_level" .= scheduledEventVoicePrivacyLevel
+      , "status" .= scheduledEventVoiceStatus
+      , "entity_type" .= Number 2
+      , "entity_id" .= scheduledEventVoiceEntityId
+      , "entity_metadata" .= Null
+      , "creator" .= scheduledEventVoiceCreator
+      , "user_count" .= scheduledEventVoiceUserCount
+      , "image" .= scheduledEventVoiceImage
+      ]
+
+data ScheduledEventPrivacyLevel = ScheduledEventPrivacyLevelGuildOnly
+
+instance ToJSON ScheduledEventPrivacyLevel where
+  toJSON ScheduledEventPrivacyLevelGuildOnly = Number 2
+
+data ScheduledEventStatus
+  = ScheduledEventStatusScheduled
+  | ScheduledEventStatusActive
+  | ScheduledEventStatusCompleted
+  | ScheduledEventStatusCancelled
+
+instance ToJSON ScheduledEventStatus where
+  toJSON ScheduledEventStatusScheduled = Number 1
+  toJSON ScheduledEventStatusActive    = Number 2
+  toJSON ScheduledEventStatusCompleted = Number 3
+  toJSON ScheduledEventStatusCancelled = Number 4
```

Comment:
Good case for the InternalDiscordEnum typeclass here [defined in the types prelude](https://github.com/aquarial/discord-haskell/blob/af2e106aaa4233e4df79a84911c3d07c8d386f7a/src/Discord/Internal/Types/Prelude.hs#L85)

Allows defining ToJSON and FromJSON

Something following the other example like

```haskell
instance InternalDiscordEnum MessageType where
  discordTypeStartValue = MessageTypeDefault
  fromDiscordType MessageTypeDefault = 0
  fromDiscordType MessageTypeRecipientAdd = 1
  fromDiscordType MessageTypeRecipientRemove = 2
  fromDiscordType MessageTypeCall = 3
  fromDiscordType MessageTypeChannelNameChange = 4
  fromDiscordType MessageTypeChannelIconChange = 5
  ....
  
instance ToJSON MessageType where
  toJSON = toJSON . fromDiscordType

instance FromJSON MessageType where
  parseJSON = discordTypeParseJSON "MessageType"
```

Reply:
changed in the following commit